clusterloader/doc/vision.md

```diff
@@ -0,0 +1,369 @@
+	// If null, object is assumed to be cluster scope.
+	NamespaceRange *Range
+	// Type definition for a given object.
+	// TODO(wojtekt): It may not be needed for Create, but
+	// it needs to be defined for update and delete anyway,
+	// so additional tax of having it here is low.
+	ObjectType ObjectType
+	// A file path to object definition.
+	ObjectTemplate string
+	// Number of instances of a given object to create per
+	// namespace.
+	ReplicasPerNamespace int32
+	// Name of TunningSet to be used.
+	TuningSet string
+}
+```
+
+This basically describes than an object (defined in a file given by its path) should
+be created in a given number of replicas in each of the namespace defined by namespace
+range and those creations should be spread over time using a referenced tuning set.
+
+
+#### PhaseUpdate
+
+PhaseUpdate defines a set of objects and a way how those should be updated within a
+single phase. We define it as following:
+
+```
+struct PhaseUpdate {
+	// Set of namespaces in which objects should be updated.
+	// If null, object is assumed to be cluster scope.
+	NamespaceRange Range
+	// Type of objects that should be updated.
+	ObjectType ObjectType
+	// Selector for objects to be updated within namespace.
+	Selector map[string]string
+	// Update function to be used to update an object.
+	UpdateFunction UpdateFunction
+	// Name of TunningSet to be used.
+	TuningSet string
+}
+```
+
+This describes that objects of a given type from a given namespaces selected by a
+given selector should be updated with a given update function and those updates
+should be spread over time using a referenced tuning set.
+
+What is still missing is defining how to do an update. As a starting point, we allow
+for two types of updates: scaling the resource (that is possible only for some
+resources exposing /scale subresource) and changing some field to given values. We
+define it as following:
+
+```
+struct UpdateFunction {
+	// Exactly one of the fields should be set.
+
+	// Scale resource using its /scale subresource
+	Scale *ScaleParameters
+	// Change a set of fields for the object to given values.
+	UpdateFields *UpdateFieldsParameters
+}
+
+struct ScaleParameters {
+	// Exactly one of the fields should be set.
+
+	// Scale to a given value.
+	DesiredReplicas *int32
+	// Scale to some random value in a given range.
+	RandomReplicas *Range
+}
+
+struct UpdateFieldsParameters {
+	UpdateFields []UpdateField
+}
+
+struct UpdateField {
+	// Path a a field that should be modified (in the same format as
+	// fields are defined in Kubernetes field selectors).
+	Field string
+	// Desired value of the field (json-serialized for non-string fields)
+	Value string
+}
+```
+
+In there will be a need to support other types of updates in the future, it should
+be fairly simple to extend these types.
+
+
+#### PhaseDelete
+
+PhaseDelete defines a set of objects that should be deleted within a single phase.
+We define it as following:
+
+```
+struct PhaseDelete {
```

Comment:
Is there some way to simply clean up all objects created?

Reply:
Not now at least. For now the easiest way would be to:
- for every object type, create a phase delete selecting all pods (label.Everything()) from all namespace (by appropriate range)

I don't think it's very painful. WDYT?

Reply:
The problem would be if we have many object types the delete phase would be rather large.

Reply:
We may consider making ObjectType to be slice of types.

Would you be ok with that?

Reply:
After thinking about it a bit more (and changing the proposal a bit to make it declarative), I think this is no longer a problem.
1. Delete is pretty declarative now.
2. If you don't care about deletes as a test owner, you may simply don't care about it.
ClusterLoader itself at the end (as "AfterAllSteps" step) will be deleting all namespaces and that would automatically delete all objects from them too.

I made that explicit in the doc.